### PR TITLE
dot,axpby,spmv: First round selecting execspace for small sizes

### PR DESCRIPTION
**Pull request by ndellingwood**
If problem size is smaller than a certain threshold use Serial
execution space.

### CODE REVIEW

**Comment by ndellingwood** [comment]
WIP PR, based on afternoon meeting with @crtrott to test performance differences of OpenMP vs Serial execution of dot, axpby, and spmv on a range of problem sizes; determined a rough estimate of thresholds to use for selecting Serial vs OpenMP for best performance. This initial commit is to demonstrate performance with prototype of this capability.

**Comment by crtrott** [comment]
This looks good, next step is collecting performance data with Tpetra CG Solve on multiple nodes.

**Comment by ndellingwood** [comment]
Reopened, made a mistake when cleaning up my messy github branches.

**Comment by ndellingwood** [comment]
Running spot-checks on white and bowman.

**Comment by ndellingwood** [comment]
This is passing all tests on white now and looking good on bowman (which is taking along time).

**Comment by ndellingwood** [comment]
White cuda tests now pass:
```
bash-4.2$ ../../scripts/test_all_sandia cuda --skip-hwloc --spot-check
Running on machine: white
Going to test compilers:  cuda/9.2.88
Testing compiler cuda/9.2.88
  Starting job cuda-9.2.88-Cuda_OpenMP-release
  PASSED cuda-9.2.88-Cuda_OpenMP-release
  Starting job cuda-9.2.88-Cuda_Serial-release
  PASSED cuda-9.2.88-Cuda_Serial-release
#######################################################
PASSED TESTS
#######################################################
cuda-9.2.88-Cuda_OpenMP-release build_time=848 run_time=599
cuda-9.2.88-Cuda_Serial-release build_time=805 run_time=998
#######################################################
FAILED TESTS
#######################################################
```

**Comment by ndellingwood** [comment]
Unit tests pass on Bowman:
```
bash-4.2$ ../../scripts/test_all_sandia intel/17 intel/18 --skip-hwloc --spot-check
Running on machine: bowman
Going to test compilers:  intel/17.2.174 intel/18.2.199
Testing compiler intel/17.2.174
  Starting job intel-17.2.174-OpenMP-release
  Starting job intel-17.2.174-Pthread-release
  PASSED intel-17.2.174-OpenMP-release
  Starting job intel-17.2.174-Serial-release
  PASSED intel-17.2.174-Pthread-release
  Starting job intel-17.2.174-OpenMP_Serial-release
  PASSED intel-17.2.174-Serial-release
Testing compiler intel/18.2.199
  Starting job intel-17.2.174-Pthread_Serial-release
  PASSED intel-17.2.174-OpenMP_Serial-release
  Starting job intel-18.2.199-OpenMP-release
  PASSED intel-18.2.199-OpenMP-release
  Starting job intel-18.2.199-Pthread-release
  PASSED intel-17.2.174-Pthread_Serial-release
  Starting job intel-18.2.199-Serial-release
  PASSED intel-18.2.199-Pthread-release
  Starting job intel-18.2.199-OpenMP_Serial-release
  PASSED intel-18.2.199-Serial-release
  Starting job intel-18.2.199-Pthread_Serial-release
  PASSED intel-18.2.199-OpenMP_Serial-release
  PASSED intel-18.2.199-Pthread_Serial-release
#######################################################
PASSED TESTS
#######################################################
intel-17.2.174-OpenMP-release build_time=1548 run_time=840
intel-17.2.174-OpenMP_Serial-release build_time=1952 run_time=3095
intel-17.2.174-Pthread-release build_time=1268 run_time=2239
intel-17.2.174-Pthread_Serial-release build_time=1693 run_time=4403
intel-17.2.174-Serial-release build_time=1111 run_time=2298
intel-18.2.199-OpenMP-release build_time=1250 run_time=820
intel-18.2.199-OpenMP_Serial-release build_time=1946 run_time=2878
intel-18.2.199-Pthread-release build_time=1031 run_time=2132
intel-18.2.199-Pthread_Serial-release build_time=1638 run_time=4090
intel-18.2.199-Serial-release build_time=972 run_time=2141
#######################################################
FAILED TESTS
#######################################################
```

**Comment by ndellingwood** [comment]
Removed the define of the `KOKKOSKERNELS_EXPERIMENTAL_ENABLE_SERIAL_LIMITS` macro in each file where it is used, this can be set from in the CMAKE_CXX_FLAGS for example. I think this is ready to merge.

**Comment by ndellingwood** [comment]
@crtrott @srajama1 could one of you review this PR? It passes test_all_sandia on bowman and white, should be ready for merge.

**Comment by ndellingwood** [comment]
@kyungjoo-kim do you have time to review this PR? I worked on it with Christian, made some fixes and it passes checks on White and Bowman.

**Comment by kyungjoo-kim** [comment]
Yes I can see the code. Did it also pass complex tests ? Also make sure that the tests generate random complex numbers (not complex numbers with zero imaginary).

**Comment by ndellingwood** [comment]
@kyungjoo-kim I didn't run any complex tests, the code changes were only for changing the execution space to serial when the problem size is smaller than a certain threshold when user sets a variable at compile time, otherwise it is the original code. I don't think complex cases should be affected unless there are pre-existing problems.

**Comment by kyungjoo-kim** [comment]
@ndellingwood Yes, the changes are not necessary for us to test complex. As we don't have complex test coverage in both KK and Trilinos, I just recommend that complex test is included in our sanity check (see trilinos/Trilinos#3493). Eventually, we will have complex coverage but it would be better if we can figure out that there is a problem in complex if there indeed is.

**Comment by ndellingwood** [comment]
@kyungjoo-kim I'll run extra complex testing sanity check, also talked with Christian about setting up a complex Jenkins test and I'll work on updating the test_all_sandia script to support that soon.

File: src/blas/KokkosBlas1_axpby.hpp [hunk 1]
```diff
@@ -87,6 +89,41 @@ axpby (const AV& a, const XMV& X, const BV& b, const YMV& Y)
   // Create unmanaged versions of the input Views.  XMV and YMV may be
   // rank 1 or rank 2.  AV and BV may be either rank-1 Views, or
   // scalar values.
+
+  #ifdef KOKKOSKERNELS_EXPERIMENTAL_ENABLE_SERIAL_LIMITS
```
[the remark this answers is not in the document]
Comment (ndellingwood) [reply src/blas/KokkosBlas1_axpby.hpp]: I talked to Christian about it, he'd like this to become an option through environment variables, so until we get that sorted out I'd rather hold off on putting it into the cmake system.

File: src/common/KokkosKernels_Serial_Limits.hpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+template < typename iType >
+struct ThresholdSizes {
+
+  static_assert( std::is_integral<iType>::value, "KokkosKernels::ThresholdSizes Error: Must be templated on integral-type" );
+
+  static constexpr iType dot_serial_limit = 20000;
+  static constexpr iType spmv_serial_limit = 200;
+  static constexpr iType axpby_serial_limit = 2000;
+
```
[the remark this answers is not in the document]
Comment (ndellingwood) [reply src/common/KokkosKernels_Serial_Limits.hpp]: This is a good idea, but I'm not sure if that would end up conflicting with Christian's idea for setting these things through environment variables.

File: src/common/KokkosKernels_Serial_Limits.hpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+#endif
+
+template <>
+struct GetSmallProblemDeviceType< Kokkos::HostSpace >
+{
+  typedef Kokkos::Device< Kokkos::Serial, Kokkos::HostSpace > type;
+};
+
+#endif
+
```
[the remark this answers is not in the document]
Comment (ndellingwood) [reply src/common/KokkosKernels_Serial_Limits.hpp]: It just falls through for the Cuda case

File: src/blas/KokkosBlas1_axpby.hpp [hunk 4]
```diff
@@ -115,6 +152,8 @@ axpby (const AV& a, const XMV& X, const BV& b, const YMV& Y)
 
   Impl::Axpby<AV_Internal, XMV_Internal, BV_Internal,
     YMV_Internal>::axpby (a_internal, X_internal, b_internal, Y_internal);
+  }
+#undef KOKKOSKERNELS_EXPERIMENTAL_ENABLE_SERIAL_LIMITS
```
[the remark this answers is not in the document]
Comment (ndellingwood) [reply src/blas/KokkosBlas1_axpby.hpp]: No I need to clean this up, good catch!
